contentcuration/contentcuration/viewsets/channel.py

```diff
@@ -505,26 +509,72 @@ def get_queryset(self):
-            .values_list("id", flat=True)
-            .distinct()
-        )
-        file_query = (
-            File.objects.filter(contentnode__tree_id=OuterRef("main_tree__tree_id"))
-            .values("checksum", "file_size")
-            .distinct()
-        )
-        queryset = queryset.annotate(
-            editors_count=SQCount(editor_query, field="id"),
-            viewers_count=SQCount(viewers_query, field="id"),
-            size=SQSum(file_query, field="file_size"),
-        )
-
         return queryset
+
+    def consolidate(self, items, queryset):
+        if items:
+            for item_channel in items:
+                metadata = self.get_or_cache_channel_metadata(
+                    item_channel["id"], item_channel["main_tree__tree_id"]
+                )
+                if metadata == DEFERRED_FLAG:
+                    item_channel["size"] = item_channel["editors_count"] = item_channel[
+                        "viewers_count"
+                    ] = DEFERRED_FLAG
+                else:
+                    item_channel.update(metadata)
+        return items
+
+    def get_or_cache_channel_metadata(self, channel_id, tree_id):
+        """
+        Returns cached data for the channel, if it exists
+        If there's not a key for the channel in the cache
+        it triggers an async task to calculate it
+        """
+        key = CACHE_CHANNEL_KEY.format(channel_id)
+        cached_info = cache.get(key)
+        if cached_info is None:
+            cache_channel_metadata_task.delay(channel_id, tree_id)
+            # from contentcuration.utils.channel import cache_channel_metadata
+            # cache_channel_metadata(channel, tree_id)
+            return DEFERRED_FLAG
+        else:
+            if "METADATA" in cached_info:
+                # do we need to add a stale strategy here or keys
+                # will be invalidated when the channel of its nodes changes
+                # or just after they expire?
+                return cached_info["METADATA"]
+            else:
+                return DEFERRED_FLAG
+
+    def get_metadata_for_channel(self, channel_id):
+        """
+        Returns cached metadata for the channel but
+        does not trigger a new task to update it if
+        there's nothing in the cache
+        """
+        key = CACHE_CHANNEL_KEY.format(channel_id)
+        cached_info = cache.get(key)
+        metadata = {
+            "size": DEFERRED_FLAG,
+            "editors_count": DEFERRED_FLAG,
+            "viewers_count": DEFERRED_FLAG,
+        }
+        if cached_info is not None:
+            if "METADATA" in cached_info:
+                return cached_info["METADATA"]
+        return metadata
+
+    @action(detail=False, methods=["get"])
+    def deferred_data(self, request):
+        ids = request.GET.get("id__in")
+        if not ids:
+            raise ValidationError("id__in GET parameter is required")
+        ids = ids.split(",")
```

Comment:
Not necessary for the admin viewset, but if we generalize this we should probably do a quick filter here by the channels the user has view permissions for, so something like:
```
ids = self.get_queryset().filter(id__in=ids).values_list("id", flat=True)
```
Will then filter down the id list to the ones that the user has permissions for (in the admin use case, this is less important though).